File: 35-Search insert Position/search_insert_position.dart
main() {
  final nums = [3, 2, 2, 3];
  final val = 3;

  final result = SearchInsertPosition.searchInsert(nums, val);
  print(result);
}

abstract class SearchInsertPosition {
  static int searchInsert(List<int> nums, int target) {
    //
    // First attempt Jun 01 2024 11:51
    // Runtime: 331ms, Memory: 150.54MB
    //
    int start = 0;
    int end = nums.length - 1;

    while (start <= end) {
      final med = start + (end - start) ~/ 2;

      if (target == nums[med]) {
        return med;
      }

      if (target > nums[med]) {
        start = med + 1;
      } else {
        end = med - 1;
      }
    }

    return start;
  }
}

File: 35-Search insert Position/SearchInsertPosition.kt
fun main() {
    val nums = intArrayOf(1, 3, 5, 6)
    val result = SearchInsertPosition.searchInsert(nums, 5)

    println(result)
}

abstract class SearchInsertPosition {
    companion object {
        fun searchInsert(nums: IntArray, target: Int): Int {
            //
            // First attempt Mar 29, 2024 04:48
            // Runtime 144ms, Memory 37.63MB
            //
            var start = 0
            var end = nums.size - 1

            while (start <= end) {
                val med = start + (end - start) / 2

                if (target == nums[med]) {
                    return med
                }

                if (target > nums[med]) {
                    start = med + 1
                } else {
                    end = med - 1
                }
            }

            return start
        }
    }
}
